### DISCUSSION TITLE
Isn't getQueryData/getQueryState reactive?

### DISCUSSION THREAD

**Discussion by chandru89new** [TanStack/query#2523] [Q&A]
I've been using getQueryData/getQueryState to fetch cache in a Nextjs project and they pretty much work like reactive elements. So when cache is updated, the components using getQueryData/State update with the latest.

However, I tried to replicate the same thing with a simple one-page codebase for a demo but it doesnt work.

Here's the code: https://codesandbox.io/s/gracious-lake-l0ddr

I am trying to load todos and "view" at almost the beginning of the app through <InitApp /> component and then trying to fetch the cached data of todos and view through getQueryData. But this doesnt work at all - even though ReactQuery Dev Tools indicates that the local cache of todos and view is updated.

I wonder what I'm doing wrong here and why in the Nextjs project, getQueryData seems to work like a reactive thing.

**Comment by TkDodo** [comment]
> Isn't getQueryData/getQueryState reactive?

no, they don't create a subscription or re-render your component, they will just give you the data "once". If you want a subscription - that is what `useQuery` is for.

**Reply by chandru89new** [reply]
Oh ok. I thought - and in my app, it does behave like this - that getQueryData/State is reactive because the component that uses it seems to get the latest data as soon as the cache is updated (even if via setQueryData).

The architecture model that I am following is derived from [re-frame][0] where you instantiate `useQuery` in some module of your codebase and then use the query name through `getQueryData` to get / subscribe to the reactive state of the data.

[0]: https://day8.github.io/re-frame/re-frame/

**Reply by TkDodo** [reply]
> Oh ok. I thought - and in my app, it does behave like this - that getQueryData/State is reactive because the component that uses it seems to get the latest data as soon as the cache is updated (even if via setQueryData).

okay, let me rephrase a bit: If you call `setQueryData`, all active observers will be informed about the update. This is the same as if data comes in from a fetch. Observers are usually created via `useQuery`. So if you call `setQueryData` and have a `useQuery` high up in the component tree, it will re-render. If you then have a child, it will re-render, and it will also get the latest data via `getQueryData`. But that child is not re-rendering because of `getQueryData`, but because the parent re-rendered.

If you use this component (the one with getQueryData) in a sub-tree that does _not_ re-render because of it's parent, because there is no `useQuery` in the parent or e.g. because there is a React.memo component somewhere that stop the re-rendering, you won't get the latest data.

So to keep components self-contained, it would be better to just `useQuery` everywhere.

**Reply by chandru89new** [reply]
Thanks for the explanation, Dominik.

**Reply by 0xyy** [reply]
@TkDodo so I am curious what is getQueryData good for, any use case ? If it's just returning not subscribed data which doesn't trigger rerender whenever new data comes or the data cache changes ?

**Reply by TkDodo** [reply]
it’s not meant to be used during rendering, as it’s not reactive, but it can be used in imperative methods and callbacks, e.g. for optimistic updates: https://tanstack.com/query/v5/docs/framework/react/guides/optimistic-updates#via-the-cache

**Reply by 0xyy** [reply]
@TkDodo this explains a lot thanks! another question, can I use useSuspenseQuery in many components like useQuery showed ? I have one main place with the useSuspenseQuery where i get the data but i want to avoid props drilling so is useSuspenseQuery in other place is ok to use ? so I will grab the data from cache that already exist not making unneecesary requests ?

**Reply by TkDodo** [reply]
Yes, there's no difference in this regard to useQuery.
